### ISSUE TITLE
Pat inject delay property 

### ISSUE THREAD

**Issue by cornae** [Patternslib/Patterns#596]
From time to time I make use of the delay property of pat-inject. Such as on the homepage of the search app of Quaive. /apps/search/

It’s also possible to build an inject with an automatic interval, which is in some cases sufficient as an alternative for real push. 

One of these cases is discussion trails. It’s not instant messaging, but it would be nice if you messages come in to a discussion that you have on your screen. For instance on the timeline. Also the timeline itself is a candidate for looping injections, so that it updates for instance once or twice a minute. Another example is counters on app icons such as the chat app or the holiday app. 

## Building loops

To build a loop you put an injection somewhere on the page (either visible or not) that has an injection statement with trigger ‘auto’ and a delay property on it. When the injection is triggered, a new injection with its own delay is delivered in the feedback so that reinfection will happen with a certain interval. It’s not atypical to have an injection target that can be updated by both a user intervention and by the auto update built with an injection with delay.

Consider the example of a discussion trail:

http://horizonscangeneesmiddelen.cornelis.amsterdam/test

A discussion trail should add new messages that are posted by other users with some delay allowed and comments posted by the current use should appear immediately. 

## Case

Try the following steps:

1. Write a message and click ‘Leave a comment’ the trail is re-injected. The trail is always capped at 4 messages. This was the five message so the oldest message is replaced by an injection link ‘Show older comments’
2. Click on ‘Show older comments’. A full injection trail is injected.
3. Wait a while and observe that both injections (the are triggers automatically in turn.

## Problem description

This means that a Pattern, in this case pat-inject is still doing some magic, even if the instance of the pattern is no longer present in the markup and therefor — at least conceptually — no longer exists. But because the counter is still running in the background, this can lead to unexpected results. Not only with loops this is a problem. Injection delays are also used to allow for animations to take place to entertain the user while content is collected and when the timer and the animation are done, the collected content is finally placed in the dom tree. Should the user in the meantime click on something else (something that doesn’t require a page refresh) that overwrites the injection trigger element, then the injection will still happen after the delay period. It could even be that the injection target no longer exists, which would result in the feedback being injected at the end of the dom tree, leading to even less predictable results.

In my opinion it would be good to kill the timer of any inject with a delay property that is no longer in the dom tree when it’s overwritten by another injection. Just like when you have a proper page refresh. @gyst indicated that this is not straight forward because the timer is tied to the window object. 

## Suggested solution
Perhaps it’s an idea to dynamically put an extra data attribute on a trigger element as soon as an injection (with delay) is triggered like ‘data-pat-inject-trigger-id=“some_generated_string”. When the timer is finished the pattern can check whether it still exists in the dom tree by scanning for this attribute with its own value. Only when it exists the injection is executed. This should prevent any accidents.

**Comment by pilz** [comment]
Very nice problem description indeed. I am learning about the delay property and how it is implemented, I wasn't really aware that we can do this already. Nice! 
I can imagine that the suggested solution could work. I will need to get my testing environment prepped first. Will do that next.

**Comment by cornae** [comment]
Thanks. The documentation was never updated other than the demo. 

The solution might be beneficial for all injections even. Maybe with an opt out possibly in the form of a property for situations i cannot think of yet.

**Comment by gyst** [comment]
I'm the author of the delay loop construct for pat-inject and Cornelis and I worked out this solution together. Let me summarize the issue and solution in my own words:

You can build a long polling loop by using pat-inject with autoload and a delay. This executes an injection only once, but if the injected snippet then has a pat-inject markup with autoload and delay, it schedules a new load and presto you have a loop as long as you keep re-injecting an autoload delayed injection. The way this works technically is a `setTimeout` which is bound to the `window` object not to the DOM element where the pat-inject is marked up on. If now *another* injection replaces the autoload-delay element before the setTimeout executes, that sets up a new setTimeout and hey you now have two loops. I worked around that in ploneintranet.messaging by some smart markup tweaking. This ticket proposes a far more robust solution:

Instead of trying to kill the timer, the timer can, when creating the timed call, mark up the injection element with a unique id; then when executing the delayed call check whether that specific id is still in the DOM and abort the injection if not.

The result is that the polling loop proceeds normally in the happy scenario. But if another injection replaces the element, the original loop will abort. Crucially, this also works if the "other injection" injects the same markup. As an example, if we'd have a loop on the messaging app count in the main navigation bar (https://github.com/quaive/ploneintranet/issues/1735) any navigation that re-injected the main navbar would fire a new loop without this fix, but it would *replace* the loop with this fix in place.

**Comment by ale-rt** [comment]
IMHO it is a bad idea to build loops like this because pat-inject wants an html response which might be costly to render and usually contains a lot of bytes that are discarded.
This is bad for the server that might be busy in rendering stuff that then is discarded (we fought with this a lot of time) and from the client point of view (in terms of battery drain and bandwidth usage, that we tend to underestimate but for the CCC user this is a huge issue).

I have been thinking quite a lot about this issue and came to a conclusion that (if we really want to use patternslib for this task) we need to take a different approach.

My vision (in case we really want to abuse `pat-inject`), is that every component that we want to update (the notification counter, the stream/comments trail, the metromap, the sidebars, the saving/review state badge, ...) should ship with a `<a class="pat-inject" data-pat-inject="event: event-name">` element and whenever `event-name` is fired the injection is performed.

The problem now is who is going to fire the event.

For the notification counter and the stream/comments trail we can definitely write a new pattern that:
1. (in a non blocking webworker) calls a dedicate view for that component in the backend (passing some parameters like the last time it was updated)
2. receives a json with a list of events it should fire

For other use cases like the metromap, sidebars, the saving/review state badge that should be updated whenever a relevant document is updated, I think there is already a parameter called `hook` that does something similar, but can fire one event at a time.

Example 1: a todo is closed in the todo sidebar, we fire an event 'todo-$UID-updated' and if in the document content we have the metromap, the metromap will reinject itself and if we have in the application window the todo view, the badge will reinject itself.
In this way we do not have to render server side the metromap if we only need the saving/review state badge, which is much cheaper to render.

Example 2: I change the document title and fire and event that causes the relevant sidebar element (if it is in the DOM: maybe we are exploring a different folder) to update itself without the need to render the whole sidebar after the form autosubmit.

I see two problems with this approach:
1. it is an hack, the world is doing that with dedicated javascript libraries/framework
2. it will not work with a static prototype

**Comment by gyst** [comment]
Good comments. Especially:
> the world is doing that with dedicated javascript libraries/framework

Doing this properly means introducing a proper UI event loop. We'll end up recreating Angular or React.

**Comment by cornae** [comment]
We should not recreate Angular or React or any other library of their kind. There are already too many of them anyway :).

I agree that some kind of push would be much better. I've been lobbying for a mechanism for a while very similar to what you suggested. I know that @pilz is also looking into it, but it hangs currently on not having the right server side component yet for handling the sockets.

This ticket however is about an improvement to inject that prevents inject from firing injections that shouldn't happen (anymore). In particular in combination with delay and more specifically when delays are used for loops (which we only do in a few place where it adds great value for the users and as long as we don't have proper push yet). 

But even without loops and even without using delays, this could be an important improvement that would still be welcome, even if we had push already.

I would be very happy if we raise a separate ticket on push, ideally preceded by a brain storming hangout.

**Comment by gyst** [comment]
This fix will only apply to delayed injections (in synchronous injections the trigger element is, by definition, present in the DOM; moreover the implementation is specific to `setTimeout`). This is, as Cornelis argues, a useful addition to the delay autoload injection we already have: it makes it far more robust.

Separating the push discussion from this seems useful. I doubt that even is in scope for Patternslib. From my developer's perspective, it may make much more sense to, in Quaive, add AngularJS to our stack and build our polling loops with that (i.e. not reinventing the wheel). But hey, we might come up with an AngularJS directive that can be exposed as a pattern. So let's have that brainstorm.
